engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+	}
+
+	// If the state has already converged, exit early.
+	stateConverged, err := serverStateConverged(obj)
+	if err != nil {
+		return checkOK, errwrap.Wrapf(err, "confirmStateConvergence failed")
+	}
+	if stateConverged {
+		return checkOK, nil
+	}
+
+	// Request changes if the server is not in the intended state yet.
+	// TODO: add permissions? (poweron, poweroff, delete)
+	checkOK = false
+	if !apply {
+		return false, nil
+	}
+	switch obj.State {
+	case StateRunning:
+		if obj.init.Debug {
+			obj.init.Logf("requesting server poweron")
+		}
+		if _, _, err := obj.client.Server.Poweron(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "failed to power server on")
+		}
+	case StateOff:
+		if obj.init.Debug {
+			obj.init.Logf("requesting server poweroff")
+		}
+		if _, _, err := obj.client.Server.Poweroff(ctx, obj.server); err != nil {
+			return false, errwrap.Wrapf(err, "failed to power server off")
+		}
+	case StateAbsent:
+		switch obj.AllowDeletion {
+		case "error":
+			return checkOK, fmt.Errorf("server deletion not allowed")
+		case "ignore":
+			// exit without waiting for convergence
+			return checkOK, nil
+		case "ifneeded":
+			if obj.init.Debug {
+				obj.init.Logf("requesting server deletion")
+			}
+			if _, err := obj.client.Server.Delete(ctx, obj.server); err != nil {
+				return false, errwrap.Wrapf(err, "failed to delete server")
+			}
+		}
+	}
+	// Wait until the requested changes are confirmed, error otherwise.
+	if err := obj.waitFor(ctx, serverStateConverged); err != nil {
+		return false, errwrap.Wrapf(err, "waitFor(serverStateConverged) exited early")
+	}
+	return checkOK, nil
+}
+
+// checkApplyServerRebuild rebuilds the server instance as needed, to ensure
+// that the resource matches the provided specifications. AllowDeletion provides
+// basic protection against unwanted server deletions.
+// TODO: consider using client.server.rebuild(image) for faster image rebuild.
+func (obj *HetznerVMRes) checkApplyServerRebuild(ctx context.Context, apply bool) (bool, error) {
+	if obj.init.Debug {
+		obj.init.Logf("checkApplyServerRebuild(apply: %t)", apply)
+	}
+	specsMatch, err := obj.cmpServerSpecs()
+	if err != nil {
+		return false, errwrap.Wrapf(err, "cmpServerSpecs failed")
+	}
+	checkOK := true
+	if specsMatch {
+		return true, nil
+	}
+	checkOK = false
+	if !apply {
+		return false, nil
+	}
+	switch obj.AllowDeletion {
+	case "error":
+		return checkOK, fmt.Errorf("server rebuild not allowed")
+	case "ignore":
+		// exit with noop
+		return checkOK, nil
+	case "ifneeded":
+		if obj.init.Debug {
+			obj.init.Logf("requesting server rebuild")
+		}
+		if err := obj.rebuildServer(ctx); err != nil {
+			return false, errwrap.Wrapf(err, "rebuildServer failed")
+		}
+	}
+	return checkOK, nil
+}
+
+// checkApplyRescueMode enables/disables rescue mode for the given keyset, to
+// ensure that rescue mode is available over SSH when needed.
+// NOTE: changing image type currently requires a manual re-enable.
```

Comment:
By "manual re-enable" I mean that you have to set the rescuemode param to `""` (disabled), run that mcl file, and then set a different image type in that param and run it again. That is currently required when you have enabled rescue mode with some image type, and want to switch to a different image type. 

I skipped that TODO earlier because in most use cases it wouldn't be an issue (in favor of other priorities), but I agree that it shouldn't become normal to ask the user to toggle any params manually to reset this property. 

It is probably best to split `checkApplyRescueMode` into two steps, something like `checkApplyRescueModeEnable` and `checkApplyRescueModeImage`. Adds another bit of complexity, but it's very similar to other checkApply steps so I don't expect too many issues.

Reply:
I added some logic to fix this issue, but I'll review it again later.